src/lib.rs

```diff
@@ -134,13 +137,130 @@
 //! }
 //! ```
 
+#![cfg_attr(not(feature = "std"), no_std)]
+
 extern crate alloc;
 
 use alloc::boxed::Box;
+use alloc::rc::{Rc, Weak as RcWeak};
+use alloc::vec::Vec;
 use core::fmt;
 use core::mem::{self, MaybeUninit};
 use core::ops::{Deref, DerefMut};
-use core::ptr;
+use core::ptr::{self, NonNull};
+
+#[cfg(feature = "std")]
+mod std_extras;
+
+/// Marker trait to indicate which types are safe to stow.
+///
+/// Currently, it is undefined behavior to read/write uninitialized memory
+/// in a primitive type, such as a pointer or usize, even if that memory is
+/// never "used". This means that types that have uninitialized bytes in their
+/// representation (such as structs with padding bytes) cannot be packed by
+/// stowaway, even if they would otherwise fit.
+///
+/// It's not currently possible to detect if a struct might contain
+/// uninitialized bytes, so any types that want to be stowed must
+/// implement this trait. MUST_BOX **must** be set to true for types which
+/// may contain uninitialized bytes; these types are still compatible with
+/// stowaway, but will be unconditionally boxed, to ensure those bytes are not
+/// packed.
+///
+/// # Safety
+///
+/// Correct usage of this trait is critical to stowaway's safety guarantees.
+/// This struct can only be implemented with MUST_BOX == false for types that
+/// can never have uninitialized bytes *anywhere* in their representation
+/// (including padding bytes).
+pub unsafe trait Stowable: Sized {
+    /// For objects with padding bytes or other uninitialized bytes that you
+    /// still want to use with stowaway, set this to true. It *is* safe to
+    /// implement `Stowable` for *any* type if MUST_BOX is true.
+    const MUST_BOX: bool = false;
+}
+
+// TODO: determine when enums are safe to stow. For instance, is an
+// Option<usize> safe to stow? if the representation is IIIIIIII?UUUUUUU,
+// where I is the int part and ? is the discriminant, probably not.
+// At the very least, types where sizeof(Option<T>) == sizeof(T) are always
+// safe to stow as long as T is safe to stow.
+
+/// Implement stowable for primitive types
+macro_rules! stowable_primitive {
+    ($($type:ty)*) => {$(
+        unsafe impl Stowable for $type {}
+    )*};
+}
+
+stowable_primitive! {
+    u8 u16 u32 u64 // u128
+    i8 i16 i32 i64 // i128
+    usize isize
+    bool char
+}
+
+#[rustversion::since(1.26)]
+stowable_primitive! {
+    u128 i128
+}
+
+/// Implement stowable for pointer-wrapper types
+macro_rules! stowable_ptr {
+    (impl <$generic:ident> for $($type:ty)*) => {$(
+        unsafe impl<$generic: ?Sized> Stowable for $type {}
+    )*}
+}
+
+stowable_ptr! {
+    impl<T> for
+    Box<T> Option<Box<T>>
+    Rc<T> Option<Rc<T>>
+    RcWeak<T> Option<RcWeak<T>>
+    NonNull<T> Option<NonNull<T>>
+}
```

Comment:
Why isn't `Arc` here? Also you can use a feature flag `alloc` to enable the `alloc` crate. This will allow you to use this crate on `no_std`.

The way I like to handle this is by doing:

```rust
#[cfg(all(not(feature = "std"), feature = "alloc"))]
extern crate alloc as std;
```
Then change all imports to use `core::` instead of `std::` except where `std` in necessary (for allocations or for `std` specific stuff)
You can also then gate all allocation related stuff on 
```rust
#[cfg(any(feature = "std", feature = "alloc"))]
```
But maybe for another PR.

Reply:
Oh, for some reason I thought that `Arc` was std-only, because of atomic operations.

All the stuff you mentioned about an std feature is in this crate already, though I can get rid of it since right now `Arc` is the only thing it was being used for.

Reply:
Also, we can't put `alloc` behind a feature flag, because the standard operation of the library includes boxing larger types. We have to use it unconditionally.